### ISSUE TITLE
[GL] InvalidOperation in SetBlendState

### ISSUE THREAD

**Issue by kvark** [gfx-rs/gfx#952]
Reported by @nullxDEADBEEF , copied the code here:
```rust
extern crate piston_window; 
extern crate cgmath;

use cgmath::{EuclideanVector, Vector};
use std::ops::*;

type Vec2 = cgmath::Vector2<f32>;
type Pos2 = cgmath::Point2<f32>;


use piston_window::*;

trait Draw {
  fn draw( &self, w : &PistonWindow, r : RenderArgs );
}

const SPEED        : f32 = 100.0;
const ACCELERATION : f32 = 75.0;

fn to_direction( angle : f32 ) -> Vec2 {
  Vec2::new( angle.cos(), angle.sin() )
}

struct Player {
  graphics : Vec<[f64; 2]>, 
  pos      : Pos2,
  vel      : Vec2,
  angle    : f32,
}

// functions regarding the player
impl Player {
  fn new( graphics : Vec<[f64; 2]>, pos : Pos2, vel : Vec2 ) -> Player {
    Player {
      graphics : graphics,
      pos      : pos,
      vel      : vel,
      angle    : 1.34,
    }
  }

  fn update_movement( &mut self, dt : f32,  controller_input : ControllerInput ) {

    let heading =
      if controller_input.move_forward {
        Vec2::new( self.angle.cos(), self.angle.sin() )
             .mul( ACCELERATION )
      } else {
        Vec2::new( 0.0, 0.0 )
      };

    if controller_input.turn_left {
      self.angle -= 1.4 * dt;
    }

    if controller_input.turn_right {
      self.angle += 1.4 * dt;
    }

    if controller_input.shoot {
      // TODO: make the player able to shoot 
    }

    self.vel = self.vel.add( &heading.mul( dt ) );

    self.vel.y = self.vel.y.min( SPEED ).max( -SPEED );
    self.vel.x = self.vel.x.min( SPEED ).max( -SPEED );

    self.pos.y += self.vel.y * dt;  
    self.pos.x += self.vel.x * dt;
    
    
  }

  fn update_collision( &mut self, dt : f32 ) {
    const SCREEN_WIDTH : f32 = 800.0;
    const SCREEN_HEIGHT : f32 = 600.0;
    const PLAYER_SIZE : f32 = 10.0;

    // screen wrap checking
    if self.pos.x < 0.0 {
      self.pos = Pos2::new( SCREEN_WIDTH - PLAYER_SIZE, self.pos.y ); 
    }

 
    if self.pos.x > SCREEN_WIDTH - PLAYER_SIZE {
      self.pos = Pos2::new( 0.0, self.pos.y );
    }

 
    if self.pos.y < 0.0 {
      self.pos = Pos2::new( self.pos.x, SCREEN_HEIGHT - PLAYER_SIZE );
    }

    if self.pos.y > SCREEN_HEIGHT - PLAYER_SIZE {
      self.pos = Pos2::new( self.pos.x, 0.0 );
    }
  }

  fn check_player( &mut self, dt : f32 ) {
  }

  fn update( &mut self, dt : f32, controller_input : ControllerInput ) {
    self.update_movement( dt, controller_input );
    self.update_collision( dt );
    self.check_player( dt );
  }
}


// we implement that the player can be drawn
impl Draw for Player {
  fn draw( &self, w : &PistonWindow, r : RenderArgs ) {
    w.draw_2d( |ctx, gfx| {

      let transform =
        ctx.transform.trans( self.pos.x as f64, self.pos.y as f64 )      
                     .rot_rad( self.angle as f64 );
      
      polygon( [0.0, 1.0, 0.0, 1.0]
             , &self.graphics[..]
             , transform
             , gfx );

      /* USED FOR DEBUGGING
      let rotx = 10.0 * self.angle.cos() as f64;
      let roty = 10.0 * self.angle.sin() as f64;

      line( [ 1.0, 0.0, 0.0, 1.0 ]
          , 0.5
          , [ self.pos.x as f64, self.pos.y as f64
            , self.pos.x as f64 + rotx, self.pos.y as f64 + roty ]
            , ctx.transform
            , gfx );
      */
    } );
  }
}

#[derive(Clone, Copy)]
struct ControllerInput {
  move_forward : bool,
  turn_left    : bool,
  turn_right   : bool,
  shoot        : bool
}

impl ControllerInput {
  pub fn new() -> ControllerInput {
    ControllerInput { move_forward: false
                    , turn_left   : false
                    , turn_right  : false
                    , shoot       : false }
  }
}

macro_rules! key_status {
    ( $event : expr, $conin : expr, $($keyname : pat => $keyfield : ident),+ ) => {
      use piston_window::*;

      if let Some( key ) = $event.press_args() {
        match key {
          $(
            Button::Keyboard( $keyname ) => $conin.$keyfield = true,
          )+
          _ => {}
        }
      }

      if let Some( key ) = $event.release_args() {
        match key {
          $(
            Button::Keyboard( $keyname ) => $conin.$keyfield = false,
          )+
          _ => {}
        }
      }
    }
}

// TODO: fix so that they cant resize the window
// and therefore avoiding collision problems
fn main() {
  // create window
  let mut window : PistonWindow = WindowSettings::new( "Asteroidy - v0.1.0", 
                                                     ( 800, 600 ) )
    .exit_on_esc( true )
    .build()
    .unwrap_or_else( |e| { panic!( "Failed to build PistonWindow: {}", e) } );
   
   
  window.set_max_fps( 60 );
  window.set_ups( 120 );

  let poly = vec![ [-10.0, -8.0], [10.0, 0.0], [-10.0, 8.0] ];

  let mut player = Player::new(   poly
                                , Pos2::new( 100.0, 100.0 )
                                , Vec2::new( 0.0,   0.0 ) );

  // handle press/release, update and render args
  let mut controller_input = ControllerInput::new();
  
  for e in window {

    key_status!( e, controller_input
               , Key::W     => move_forward
               , Key::A     => turn_left
               , Key::D     => turn_right
               , Key::Space => shoot );

    if let Some( ua ) = e.update_args() {
      player.update( ua.dt as f32, controller_input );
    }
    
    if let Some( ra ) = e.render_args() {
      e.draw_2d( |_c, g| {
        clear( [0.0, 0.0, 0.0, 1.0], g );
      }); 

      if controller_input.shoot {
      }
      player.draw( &e, ra );
    }
  }
}
```

> The error I get when i do 'cargo run':
> thread '<main>' panicked at 'Error after executing command SetBlendState(0, Color { mask: RED | GREEN | BLUE | ALPHA | MASK_ALL, blend: Some(Blend { color: BlendChannel { equation: Add, source: ZeroPlus(SourceAlpha), destination: OneMinus(SourceAlpha) }, alpha: BlendChannel { equation: Add, source: One, destination: One }}) }): InvalidOperation', /home/agdev/.cargo/registry/src/github.com-88ac128001ac3a9a/gfx_device_gl-0.9.0/src/lib.rs:283
note: Run with `RUST_BACKTRACE=1` for a backtrace.
>Process didn't exit successfully: `target/debug/asteroidy` (exit code: 101)

**Comment by kvark** [comment]
@nullxDEADBEEF - could you make an `apitrace` capture and provide the dump?

**Comment by nullxDEADBEEF** [comment]
[asteroidy.txt](https://github.com/gfx-rs/gfx/files/225996/asteroidy.txt)

Here is the apitrace dump

**Comment by nullxDEADBEEF** [comment]
[glxinfo_asteroidy.txt](https://github.com/gfx-rs/gfx/files/226036/glxinfo_asteroidy.txt)

The glxinfo

**Comment by kvark** [comment]
@tomaka suggests that the ARB versions of the functions, while being supported, may have different entry points from the core ones (that are not supported on GL-3.2). We need to change our `glEnablei`, `glBlendEquationSeparatei`, `glBlendFuncSeparatei` and `glColorMaski` to call ARB versions, at least for a test.

**Comment by kvark** [comment]
Fixed by #953
